includes/admin/class-amp-story-templates.php

```diff
@@ -0,0 +1,197 @@
+<h1 style="font-size:28px;width:76%;height:9%;position:absolute;top:27%;left:13%" class="has-text-color has-very-light-gray-color" id="ddce50c3-5776-44c5-9d9b-0d6d4b12d88e" data-font-family="Garamond"><amp-fit-text layout="fill" class="amp-text-content">Hello, Templates!</amp-fit-text></h1>
+<!-- /wp:amp/amp-story-text --></amp-story-grid-layer></amp-story-page>
+<!-- /wp:amp/amp-story-page -->',
+			),
+		);
+	}
+
+	/**
+	 * Determine if a post exists based on title, content, only receives a published post.
+	 *
+	 * Note that this is almost a copy of the default `post_exists` function except for the post status.
+	 *
+	 * @global wpdb $wpdb WordPress database abstraction object.
+	 *
+	 * @param string $title Post title.
+	 * @param string $content Post content.
+	 * @return int Post ID if post exists, 0 otherwise.
+	 */
+	private function template_exists( $title, $content ) {
+		if ( empty( $title ) || empty( $content ) ) {
+			return 0;
+		}
+		global $wpdb;
+
+		$post_title   = wp_unslash( sanitize_post_field( 'post_title', $title, 0, 'db' ) );
+		$post_content = wp_unslash( sanitize_post_field( 'post_content', $content, 0, 'db' ) );
+
+		return (int) $wpdb->get_var(
+			$wpdb->prepare(
+				"SELECT ID FROM $wpdb->posts WHERE post_title = %s AND post_content = %s AND post_status = 'publish'",
+				array( $post_title, $post_content )
+			)
+		);
+	}
+
+	/**
+	 * Register taxonomy for differentiating AMP Story Templates.
+	 */
+	private function register_taxonomy() {
+		register_taxonomy(
+			self::TEMPLATES_TAXONOMY,
+			'wp_block',
+			array(
+				'query_var'             => self::TEMPLATES_TAXONOMY,
+				'show_admin_column'     => false,
+				'show_in_rest'          => true,
+				'rest_base'             => self::TEMPLATES_TAXONOMY . 's',
+				'rest_controller_class' => 'WP_REST_Terms_Controller',
+				'public'                => false,
+				'show_ui'               => false,
+				'show_tagcloud'         => false,
+				'show_in_quick_edit'    => false,
+				'hierarchical'          => false,
+				'show_in_menu'          => false,
+				'meta_box_cb'           => false,
+			)
+		);
+
+		if ( ! term_exists( self::TEMPLATES_TERM, self::TEMPLATES_TAXONOMY ) ) {
+			wp_insert_term(
+				__( 'Story Template', 'amp' ),
+				self::TEMPLATES_TAXONOMY,
+				array(
+					'description' => __( 'Story Template', 'amp' ),
+					'slug'        => self::TEMPLATES_TERM,
+				)
+			);
+		}
+	}
+
+	/**
+	 * Filter REST request for reusable blocks to not display templates under Reusable Blocks within other posts.
+	 *
+	 * @param array           $args Original args.
+	 * @param WP_REST_Request $request WP REST Request object.
+	 * @return array Args.
+	 */
+	public function filter_rest_wp_block_query( $args, $request ) {
+		$headers = $request->get_headers();
+		if ( ! isset( $headers['referer'][0] ) ) {
+			return $args;
+		}
+
+		$parts = wp_parse_url( $headers['referer'][0] );
```

Comment:
Checking the referrer feels a bit hacky.

What if we passed an `?amp_story=123` param in the REST API request instead? It could be achieved using an [`apiFetch` middleware](https://wordpress.org/gutenberg/handbook/designers-developers/developers/packages/packages-api-fetch/#middlewares) that does this.

Then, here we only have to check for the existence and validity of the query param.

Reply:
It does feel a bit hacky, true.

The goal of this filter is to process the request that's done in the Block Inserter on Post types which are not AMP Story Post types to prevent the templates showing up in inserters. Otherwise templates show up under Reusable Blocks and just fail when inserted.

https://github.com/WordPress/gutenberg/blob/master/packages/editor/src/store/effects/reusable-blocks.js#L64

Could you elaborate on what you meant by passing `amp_story` param in the REST API? Not sure I understand what you meant.

Reply:
Sure.

My train of thought was that when editing an AMP story, we would filter any REST API requests for reusable blocks and add a query param of our choice to the URL. For example `../../blocks?show_amp_story_templates_please=1`. That's what the middleware would do.

Then, on the PHP side, we could probably _always_ add the `tax_query` _unless_ that `show_amp_story_templates_please` query param is detected.

Reply:
Ah, okay, now I see, basically whitelisting only AMP Story instead of trying to blacklist all the others. Makes sense. This would require adjusting fetching the templates for an AMP Story instead of just using the default fetching.

There's one other moment to consider -- when the template itself is being edited (in the future). Eventually a template could consist of more than one template, and probably the default Inserter should display the Reusable blocks (templates) when editing a template, too.